intellij/org.eclipse.xtend.idea/src/org/eclipse/xtend/core/idea/highlighting/XtendHighlightingLexer.xtend

```diff
@@ -0,0 +1,189 @@
+import org.eclipse.xtext.idea.parser.AntlrDelegatingIdeaLexer
+import org.eclipse.xtext.parser.antlr.AntlrTokenDefProvider
+
+import static org.eclipse.xtend.core.idea.highlighting.XtendHighlightingLexer.*
+
+/**
+ * @author Jan Koehnlein - Initial contribution and API
+ */
+class XtendHighlightingLexer extends LexerBase {
+
+	protected static class RichTextToken {
+		int index = 0
+		int[] offsets
+		int[] lengths
+		boolean delimiter
+		IElementType tokenType
+
+		def getTokenLength() {
+			lengths.get(index)
+		}
+
+		def getTokenOffset() {
+			offsets.get(index)
+		}
+
+		def hasNext() {
+			index < offsets.length - 1
+		}
+
+		def advance() {
+			index++;
+			delimiter = !delimiter
+		}
+
+		def getTokenType() {
+			if (delimiter)
+				GUILLEMET_ELEMENT_TYPE
+			else
+				tokenType
+		}
+
+		def getType() {
+			if (delimiter)
+				GUILLEMET_ELEMENT_TYPE.index
+			else
+				tokenType.index
+		}
+	}
+
+	public static val GUILLEMET_ELEMENT_TYPE = new IElementType(XtendHighlightingStyles.RICH_TEXT_DELIMITER_ID,
+		XtendLanguage.INSTANCE)
+
+	static val TOKEN_TYPE_NAMES_CONTAINING_GUILLEMETS = #{
+		'RULE_RICH_TEXT',
+		'RULE_RICH_TEXT_START',
+		'RULE_RICH_TEXT_END',
+		'RULE_RICH_TEXT_INBETWEEN',
+		'RULE_COMMENT_RICH_TEXT_INBETWEEN',
+		'RULE_COMMENT_RICH_TEXT_END'
+	}
+
+	@Inject AntlrDelegatingIdeaLexer delegate
+
+	@Inject extension AntlrTokenDefProvider
+
+	RichTextToken _currentRichTextToken
+
+	override advance() {
+		if (currentRichTextToken?.hasNext) {
+			currentRichTextToken.advance
+		} else {
+			_currentRichTextToken = null
+			delegate.advance
+		}
+	}
+	
+	def getCurrentRichTextToken() {
+		if(_currentRichTextToken == null) {
+			if(delegate.currentToken != null) {
+				val tokenID = tokenDefMap.get(delegate.currentToken.type)
+				if (TOKEN_TYPE_NAMES_CONTAINING_GUILLEMETS.contains(tokenID))
+					_currentRichTextToken = createRichTextToken(tokenID, delegate.tokenStart,
+						delegate.tokenEnd - delegate.tokenStart, delegate.tokenType)
+			}
+		}
+		return _currentRichTextToken
+	}
+
+	override getState() {
+		if (currentRichTextToken != null)
+			currentRichTextToken.type
+		else
+			delegate.state
+	}
+
+	override getTokenEnd() {
+		if (currentRichTextToken != null)
+			currentRichTextToken.tokenOffset + currentRichTextToken.tokenLength
+		else
+			delegate.tokenEnd
+	}
+
+	override getTokenStart() {
+		if (currentRichTextToken != null)
+			currentRichTextToken.tokenOffset + currentRichTextToken.tokenLength
+		else
+			delegate.tokenEnd
+	}
+
+	override getTokenType() {
+		if (currentRichTextToken != null)
+			currentRichTextToken.getTokenType
+		else
+			delegate.tokenType
+	}
+
+	override start(CharSequence buffer, int startOffset, int endOffset, int initialState) {
+		delegate.start(buffer, startOffset, endOffset, initialState)
+	}
+
+	override getBufferEnd() {
+		delegate.bufferEnd
+	}
+
+	override getBufferSequence() {
+		delegate.bufferSequence
+	}
+
+	override restore(LexerPosition position) {
```

Comment:
This is usually used since the lexer is used incrementally for the highlighting.